osu.Game/Screens/Select/SongSelect.cs

```diff
@@ -0,0 +1,450 @@
+﻿// Copyright (c) 2007-2017 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using OpenTK;
+using OpenTK.Input;
+using osu.Framework.Allocation;
+using osu.Framework.Audio;
+using osu.Framework.Audio.Sample;
+using osu.Framework.Audio.Track;
+using osu.Framework.Configuration;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Primitives;
+using osu.Framework.Graphics.Transforms;
+using osu.Framework.Input;
+using osu.Framework.Screens;
+using osu.Framework.Threading;
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.Drawables;
+using osu.Game.Database;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Containers;
+using osu.Game.Modes;
+using osu.Game.Overlays;
+using osu.Game.Screens.Backgrounds;
+using osu.Game.Screens.Select.Options;
+
+namespace osu.Game.Screens.Select
+{
+    public abstract class SongSelect : OsuScreen
+    {
+        private Bindable<PlayMode> playMode = new Bindable<PlayMode>();
+        private BeatmapDatabase database;
+        protected override BackgroundScreen CreateBackground() => new BackgroundScreenBeatmap(Beatmap);
+
+        private CarouselContainer carousel;
+        private TrackManager trackManager;
+        private DialogOverlay dialogOverlay;
+
+        private static readonly Vector2 wedged_container_size = new Vector2(0.5f, 225);
+        private BeatmapInfoWedge beatmapInfoWedge;
+
+        private static readonly Vector2 background_blur = new Vector2(20);
+        private CancellationTokenSource initialAddSetsTask;
+
+        private SampleChannel sampleChangeDifficulty;
+        private SampleChannel sampleChangeBeatmap;
+
+        private List<BeatmapGroup> beatmapGroups;
+
+        /// <summary>
+        /// Can be null if hasFooter:false in constructor
+        /// </summary>
+        protected BeatmapOptionsOverlay BeatmapOptions { get; private set; }
+        /// <summary>
+        /// Can be null if hasFooter:false in constructor
+        /// </summary>
+        protected Footer Footer { get; private set; }
+
+        private FilterControl filter;
+        public FilterControl Filter
+        {
+            get
+            {
+                return filter;
+            }
+            private set
+            {
+                if (filter != value)
+                {
+                    filter = value;
+                    filterChanged();
+                }
+            }
+        }
+
+        protected SongSelect(bool hasFooter)

```

Comment:
i think i prefer using `protected virtual bool ShowFooter => true` rather than ctor for things like this. Allows for easier access and better visibility (and stops ctors getting absurdly huge to handle specific cases).